test/e2e/storage/pvc_storageclass.go

```diff
@@ -32,158 +32,162 @@ import (
-			framework.ExpectNoError(err, "Error updating PVC with the correct storage class")
-
-			// Create PV with specific class
-			pv := e2epv.MakePersistentVolume(e2epv.PersistentVolumeConfig{
-				NamePrefix:       "pv-",
-				StorageClassName: storageClass.Name,
-				VolumeMode:       pvc.Spec.VolumeMode,
-				PVSource: v1.PersistentVolumeSource{
-					HostPath: &v1.HostPathVolumeSource{
-						Path: "/tmp/test",
-					},
-				},
-			})
-			_, err = e2epv.CreatePV(ctx, client, f.Timeouts, pv)
-			framework.ExpectNoError(err, "Error creating pv %v", err)
-			ginkgo.DeferCleanup(e2epv.DeletePersistentVolume, client, pv.Name)
-
-			// Verify the PVC is bound and has the new default SC
-			claimNames := []string{pvc.Name}
-			err = e2epv.WaitForPersistentVolumeClaimsPhase(ctx, v1.ClaimBound, client, namespace, claimNames, 2*time.Second /* Poll */, t.Timeouts.ClaimProvisionShort, false)
-			framework.ExpectNoError(err)
-			updatedPVC, err := client.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, pvc.Name, metav1.GetOptions{})
-			framework.ExpectNoError(err)
-			gomega.Expect(*updatedPVC.Spec.StorageClassName).To(gomega.Equal(storageClass.Name), "Expected PVC %v to have StorageClass %v, but it has StorageClass %v instead", updatedPVC.Name, prefixSC, updatedPVC.Spec.StorageClassName)
-			framework.Logf("Success - PersistentVolumeClaim %s got updated retroactively with StorageClass %v", updatedPVC.Name, storageClass.Name)
-		})
+	f.It("should assign default StorageClass to PVCs retroactively", f.WithDisruptive(), f.WithSerial(), func(ctx context.Context) {
+		defaultSCs, err := getDefaultStorageClasses(ctx, client)
+		framework.ExpectNoError(err, "Failed to get default StorageClasses")
+
+		defer func() {
+			// Restore existing default StorageClasses at the end of the test
+			for _, sc := range defaultSCs {
+				setStorageClassDefault(ctx, client, sc.Name, "true")
+			}
+		}()
+
+		// Unset all default StorageClasses
+		for _, sc := range defaultSCs {
+			klog.InfoS("Unsetting default StorageClass", "StorageClass", sc.Name)
+			setStorageClassDefault(ctx, client, sc.Name, "false")
+		}
+
+		// Ensure no default StorageClasses exist
+		if len(defaultSCs) > 0 {
+			err = ensureNoDefaultStorageClasses(ctx, client)
+			framework.ExpectNoError(err, "Failed to ensure no default StorageClasses exist")
+		}
+
+		// Create a PVC with nil StorageClass and confirm it is in a Pending state
+		pvc := createPVC(ctx, client, namespace)
+		defer deletePVC(ctx, client, namespace, pvc.Name)
+		err = e2epv.WaitForPersistentVolumeClaimPhase(ctx, v1.ClaimPending, client, namespace, pvc.Name, time.Second, 30*time.Second)
+		framework.ExpectNoError(err, "PVC should be in Pending state when no default StorageClass exists")
+
+		// Create a default StorageClass
+		sc := createDefaultStorageClass(ctx, client)
+		defer deleteStorageClass(ctx, client, sc.Name)
+
+		// Create PV with the default StorageClass
+		pv := createPV(ctx, client, f, sc.Name, pvc.Spec.VolumeMode)
```

Comment:
The e2e test aims to verify the pvc assigned to a specific storage class when the storage class is not specified in the pvc and the cluster has a default storage class, so we don't need to verify the binding of the pvc and pv in this case. @RomanBednar what do you think?

For example,

```go
gomega.Eventually(ctx, framework.GetObject(f.ClientSet.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get, pvc.Name, metav1.GetOptions{})).
  WithPolling(framework.Poll).WithTimeout(framework.ClaimProvisionTimeout).Should(gomega.HaveField("Spec.StorageClassName",
  gomega.Equal(defaultSVC.Name)), "failed to wait for PVC to have the storageclass %s", defaultSVC.Name)
```

Reply:
@carlory It aims to verify SC assignment while PVC with no SC is present and *after* some SC is set as default. Anyway, what you suggest could be sufficient and test could be limited to just checking SC field on PVC I think.

Reply:
Simplified test further as per the suggestion above.